aggregator/handler/router/traffic_controlled.go

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package router
+
+import (
+	"github.com/m3db/m3aggregator/aggregator/handler/common"
+)
+
+type trafficControlledRouter struct {
+	*common.TrafficController
+	Router
+}
+
+// NewTrafficControlledRouter creates a traffic controlled router.
+func NewTrafficControlledRouter(
+	trafficController *common.TrafficController,
+	router Router,
+) Router {
+	return &trafficControlledRouter{
+		TrafficController: trafficController,
+		Router:            router,
+	}
+}
+
+func (r *trafficControlledRouter) Route(shard uint32, buffer *common.RefCountedBuffer) error {
+	if !r.TrafficController.Allow() {
+		return nil
+	}
+	return r.Router.Route(shard, buffer)
+}
+
+func (r *trafficControlledRouter) Close() {
+	r.Router.Close()
```

Comment:
This is a good opportunity to close the traffic controller if it has a `Close` method.